kernel: configure MPU only if switching to a different process

alphan opened the pull request:
### Pull Request Overview

Currently, the scheduler configures the MPU every time when it switches
to a process. This change aims to improve syscall latency by caching the
last running process and configuring the MPU only when switching to a
different process.

See also: #1730.

### Testing Strategy

I ran a simple `libtock-rs` application for measuring syscall latency with
and without this change on opentitan on verilator. This change reduces the
latency of a command syscall by ~~2751~~ 1711 cycles (the exact number of cycles
saved obviously depends on the actual implementation, e.g. would probably
be less after #1821).

Also ran `make allcheck`.

### TODO or Help Wanted

I couldn't find any doc files or tests that should be updated. Since
this is my first real Tock PR, it would be great if you could point me
in the right direction if I missed anything.

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make formatall`.

ppannuto commented on kernel/src/sched.rs:
```diff
@@ -331,6 +336,7 @@ impl Kernel {
         systick.reset();
         systick.set_timer(KERNEL_TICK_DURATION_US);
         systick.enable(false);
+        let identifier: usize = process.appid().id();
```
```suggestion
        let identifier = process.appid();
```

Ah, I see where this is a bit non-intuitive, but the plain `appid` should be used in-kernel; the `.id()` method is only for special situations: https://docs.tockos.org/kernel/struct.appid#method.id

bradjc replied:
This is subtle, but we should never make a copy of a process's ID like this, since it can change (see 4003b297c90f693ea297a3c70de35f28a0d49c7f for an example of where this has lead to a bug in the past). Instead just use `process.appid()` when needed.

alphan replied:
Thank you very much for reviewing this @ppannuto and @bradjc! Your comments raise two very important points:

**Caching process IDs should, in general, be avoided**
Thanks for pointing me to #1613 @bradjc! It is indeed a subtle bug caused by stale process IDs. However, this PR actually relies on the fact that the process IDs returned by `process.appid().id()` is unique among different processes and different _instances_ of the _same_ process [1]. This ensures that the MPU will be configured whenever we switch to a different process. I guess, another alternative could be to store this information as a flag in the process structure itself, but I find this approach more intuitive.

**`process.appid()` vs `process.appid().id()`**
@ppannuto, the reason for using `process.appid().id()` was performance. Using `process.appid()` introduces an overhead of 150 cycles.

[1]  Apologies for the abuse of terminology. Here, I am referring to processes that are restarted after crashes or updated at runtime.

ppannuto replied:
I'm afraid this does the opposite of what you might want. It is perfectly valid for the `.id()` value to be re-used. In fact, right now it is simply the index into the static array of `Process` pointers. If a process restarts, `.appid()` will change and `.appid().id()` will not.

However, if a process is restarted, its memory layout likely _has_ changed (i.e. no `brk` yet), thus the MPU regions should be reconfigured. I believe `.appid()` is still the correct choice, as it is what is unique among instances of a process.

alphan replied:
Oh, I was under the impression that `index` can be reused but `.id()`, i.e. `identifier`, is unique because `create_process_identifier()` is called when a process is created or restarted.

ppannuto replied:
Yes, sorry, you are right. I described the old behavior (id used to be the index; #1566 updated this).

Nonetheless, comparing AppId should do the right thing (it internally compares the identifiers returned by `.id()`):
```rust
impl PartialEq for AppId {
    fn eq(&self, other: &AppId) -> bool {
        self.identifier == other.identifier
    }
}
```

---

I know the _intent_ of the `AppId`/`identifier` split was to use `AppId`s internally and to provide `identifier`s for external use, but I'm not sure I could articulate when/where/why i'ts preferred to use `AppId` over the `identifer`

alphan replied:
No worries, sounds good! I will update this PR to use `AppId` (we can hopefully avoid the overhead by using `MapCell` instead of `Cell`).

bradjc commented:
We would need to update the kernel/src/platform/mpu.rs documentation, particularly to clarify that the MPU implementation must be setup to only enforce in the userspace privilege level.

bradjc commented on kernel/src/sched.rs:
```diff
@@ -47,6 +47,10 @@ pub struct Kernel {
+
+    /// Identifier of the last running process. This is used to determine if
+    /// we are swtiching to a different process and need to configure the MPU.
+    last_running_process: Cell<Option<usize>>,
```
```suggestion
    last_running_process: OptionalCell<AppId>,
```

alphan commented:
>  Did your testing include running multiple apps as well?

Yes, I also tried this with multiple apps and didn't notice any issues.

> Generally, I think this is a good change. However, I'm concerned there may be some extra considerations where we may need to identify the MPU configuration as "tainted". Specifically:
> 
> * if there are additional grant allocations, then the deny region may need to grow (though, this is likely not an immediate issue as the allowed region is generally minimal)
> * if there was a change in permitted region, i.e. the application called `brk` or an IPC `allow`* then these need to be recalculated
> * Maybe others I'm not thinking of?
> 
> Basically, I think this is a good common-case optimization, but I think we also need a way for something to "poison" the MPU configuration to force it to be recalculated.
> 
> *though that second case may actually be fine since in the current architecture as it's always a different process that is allowing a region, so we'll go through regular process switch

Thank you for catching this! I actually looked at various places that can change the MPU config but missed at least one case where MPU config can be modified and not applied immediately, e.g. `sbrk` and `brk` eventually call `configure_mpu()` but `alloc` doesn't. Since cache invalidation is tricky, does it make sense to modify `MPU` and `MPUConfig` to keep track of the _taintedness_ of `MPUConfig` at a higher level?

ppannuto commented:
>  Since cache invalidation is tricky, does it make sense to modify MPU and MPUConfig to keep track of the taintedness of MPUConfig at a higher level?

Yeah, I think that's a great idea. I believe the mpu configuration structure is already well-equipped for this as all of its state is private and can only be modified by getters and setters; hooking the latter to mark the MPU configuration as dirty, and then perhaps updating `configure_mpu` to know what process it's configured for, pushing the hot-path logic into the MPU itself seems like a good design to me.

In that case, I actually imagine we could _remove_ the `configure_mpu`s that are sprinkled around, instead architecting things such that whenever a process is returned to the MPU layer is consulted first, updating itself if necessary.

alphan commented:
- Reverted changes in `kernel/src/sched.rs`.
- Added `app_id` parameter to `configure_mpu`.
- Implemented the idea captured in this PR as a risc-v-only optimization in `arch/rv32i/src/pmp.rs`.
- Rebased on master.
- Also made `disable_mpu` a no-op for a more accurate latency measurement (moved its contents to `PMPConfig::clear_mpu_registers`, called from `configure_mpu` to stay close to current behavior). It looks like we don't need this function at all but I am deferring this change to #1860, which I created to keep track of the things that I noticed while working on this.

alphan commented:
@ppannuto, thank you for reviewing this change and your comments. I have a question regarding the changes in `arch/cortex-m3/src/mpu.rs`. Shouldn't `hardware_is_configured_for` be defined in `MPU` (to track the state of the hw)? Also, don't we need a `is_dirty` field in `CortexMConfig`?

ppannuto commented:
Hmm.., you're right we need some form of structure that is the global to the MPU hardware. The MPU configuration is stored _per process_ ([here](https://github.com/tock/tock/blob/master/kernel/src/process.rs#L789)), so doing dirty tracking in the `MpuConfig` type won't be sufficient for resetting the MPU when we change processes.

I don't think the `CortexMConfig` needs an explicit `dirty` flag. Instead, the cell indicates what application the hardware "is configured for"; whenever the MPU configuration is changed, that cell is cleared to indicate that the MPU is not configured correctly for any application, which acts similarly to a separate dirty signal.

alphan commented:
The `MPU` struct that I was referring to is the one defined [here](https://github.com/tock/tock/blob/master/arch/cortex-m3/src/mpu.rs#L124). IIUC, this should track the state of the hardware, i.e. `hardware_is_configured_for`, while `CortexMConfig` tracks if it is dirty or not.

alphan commented:
@ppannuto, 121b33f844b62930babbc2cd3d013606b14f28b6 LGTM, thanks! I also pushed a commit to fix the way `last_configured_for` is used in `configure_mpu`.

alistair23 commented on arch/rv32i/src/pmp.rs:
```diff
@@ -352,8 +354,18 @@ impl kernel::mpu::MPU for PMPConfig {
+
+    fn enable_mpu(&self) {}
+
+    fn disable_mpu(&self) {}
```
This came up in: https://github.com/tock/tock/pull/1871 and isn't the right approach as people might still want to disable PMP at boot.

alistair23 replied:
It also means we now don't match the trait documentation.

alphan replied:
I think there is some room for improvement in the trait. The doc says that this will be called before returning to the kernel, so I think it is reasonable to implement it as NOP for risc-v similar to `enable_mpu`, but the name is too general so it amy be misleading.

alistair23 replied:
There is a pending PR to update the trait which will then stop calling disable on every transition.

alphan commented on arch/rv32i/src/pmp.rs:
```diff
@@ -478,148 +492,164 @@ impl kernel::mpu::MPU for PMPConfig {
+            });
+
+            // tock/tock#1860
+            self.clear_mpu_registers();
```
As I mentioned in #1860, I don't think we need to clear the PMP registers when returning to the kernel since we don't use the lock bit. Calling this function here just to stay as close as possible to the current implementation, was planning to remove it as part of #1860.

alistair23 replied:
This is confusing then. Can we leave the current `disable_mpu()` as is and work on fixing that? There is already a PR for this: https://github.com/tock/tock/pull/1873

alphan commented:
Reverted changes in `disable_mpu` to address @alistair23's comments. We should see the expected latency improvements when `do_process` stops clearing PMP registers every time it returns to kernel (#1873).

alistair23 commented on arch/rv32i/src/pmp.rs:
```diff
@@ -478,148 +490,161 @@ impl kernel::mpu::MPU for PMPConfig {
+        // Skip PMP configuration if it is already configured for this app and the MPU
+        // configuration of this app has not changed.
+        if !last_configured_for_this_app || config.is_dirty.get() {
+            // tock/tock#1860
```
Can you remove this?

alphan replied:
Hmm, I thought the comment and the PR description were pretty clear. This skips PMP configuration if it is already configured for the given app and has not changed since the last time it was committed. Please see also #1730 and #1860.

alistair23 replied:
The comment does that?

alphan replied:
Just to be clear, are you referring to just `tock/tock#1860` or the `if` statement? It is hard to tell looking at the way your comment is rendered on github (my reply above is for the latter but probably you meant the former). I added a reference to #1860 because the regions are sorted every time PMP is configured.

alistair23 replied:
Yeah, I was referring to the comment. The GitHub UI is really lacking unfortunately, so it can be confusing.

Can we remove the comment then? I don't think it adds anything and we have a tracking issue to track the inefficiently.

alphan replied:
Ah, apologies for the confusion (I know github shows some context above the line that you comment but I was confused because this one is perfectly aligned). Done.

alistair23 replied:
Thanks! It's good to have someone else looking at the PMP code :)

silvestrst replied:
I think it might be useful to refer to the Github issue in the PR description, and instead add a descriptive comment here.

silvestrst commented on arch/rv32i/src/pmp.rs:
```diff
@@ -99,10 +102,11 @@ impl PMPRegion {
 pub struct PMPConfig {
     regions: [PMPRegion; 8],
     total_regions: usize,
+    is_dirty: Cell<bool>,
+    last_configured_for: MapCell<AppId>,
```
Could be useful to document these fields.

alphan replied:
Done.

silvestrst commented on arch/rv32i/src/pmp.rs:
```diff
@@ -154,6 +160,8 @@ impl PMPConfig {
             // As we use the PMP TOR setup we only support half the number
             // of regions as hardware supports
             total_regions: pmp_regions / 2,
+            is_dirty: Cell::new(true),
+            last_configured_for: MapCell::empty(),
```
Could you maybe drop an empty line before `is_dirty`, otherwise it looks like the comment applies to the block of code and not just `is_dirty`.

alphan replied:
Done.

silvestrst commented on arch/rv32i/src/pmp.rs:
```diff
@@ -478,148 +492,161 @@ impl kernel::mpu::MPU for PMPConfig {
+                    None => {}
+                };
+            }
+            // Clear the dirty flag
```
Non informative comment, it is obvious from the code that the dirty flag is cleared.

alphan replied:
Done.

silvestrst commented on arch/rv32i/src/pmp.rs:
```diff
@@ -478,148 +492,161 @@ impl kernel::mpu::MPU for PMPConfig {
-                            csr::CSR.pmpcfg3.set(cfg_val << 24 | csr::CSR.pmpcfg0.get());
-                            csr::CSR.pmpaddr15.set((start as u32 + size as u32) >> 2);
+    fn configure_mpu(&self, config: &Self::MpuConfig, app_id: &AppId) {
+        self.last_configured_for.map(|last_app_id| {
```
I think it could be useful to document what this block does.

alphan replied:
Latest version should be clearer but added a short comment nevertheless.

silvestrst commented on arch/rv32i/src/pmp.rs:
```diff
@@ -478,148 +492,161 @@ impl kernel::mpu::MPU for PMPConfig {
+                return;
+            }
+            // tock/tock#1860
+            let mut regions_sorted = config.regions.clone();
```
I think it could be useful to document what this block does.

alphan replied:
This part was already there and will hopefully be removed, added a brief comment.

alphan commented:
Thank you for reviewing this @bradjc and @silvestrst!

Added comments to address @silvestrst's comments.

ppannuto commented:
bors r+

bradjc commented:
big queue i guess?
